blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+```ts
+export interface Book {
+  ...
+  published: number;
+}
+
+export interface BookView {
+  ...
+  published: Date;
+  isNew(): boolean;
+}
+
+export const calculateNew = (book: Book) => {
+  return new Date().getFullYear() === new Date(book.published).getFullYear();
+};
+
+export const getBookView = () =>
+  createSelector(
+    ...
+    (book, authors, tags): BookView =>
+      book && {
+        ...
+        published: new Date(book.published),
+        isNew: () => calculateNew(book),
+      }
+  );
+```
+
+The result can be observed in the next [StackBlitz Example](https://stackblitz.com/github/dhhyi/ngrx-data-views/tree/data-views-as-business-objects?file=src%2Fapp%2Fstore%2Fbook-view%2Findex.ts&devtoolsheight=50). As you can see in the console, solving it this way leads to many calls to the "new" method provided by the data view. The reason lies within the Angular change detection.
+
+## Memoized Data View Methods
+
+When developing, you are composing all components living in an Angular application into a tree for rendering. Likewise Angular keeps a tree of corresponding change detectors for optimizing view updates. All bindings in templates referencing component properties are checked in every change detector cycle. But when one is using call expressions on the template (which we would be doing when accessing `book.isNew()`), the binding cannot be properly cached and thus has to be re-evaluated in every cycle.
+
+There are multiple ways of tackling this issue:
+
+1. Cache the results of call expressions:
+
+   In this easy example you could just calculate the result once and keep it as a simple property on the `BookView` instance instead. Everything in the store will stay properly normalized, it is just the data view supplying derived data for easy access:
+
+   ```ts
+   export interface BookView {
+     ...
+     isNew: boolean;
+   }
+
+   export const getBookView = () =>
+     createSelector(
+       ...
+       (book, authors, tags): BookView =>
+         book && {
+           ...
+           isNew: calculateNew(book),
+         }
+     );
+   ```
+
+   In other cases the `ngOnChanges` in component classes could be used to cache expensive calculations to fields. And then those fields could be bound to on the templates.
+
+2. Disable Change Detection:
+
+   By choosing `OnPush` [change detection](https://angular.io/api/core/ChangeDetectionStrategy), the standard change detection triggered by every browser event can be switched off and change detection is only run when `@Input` decorated fields of a component are changing. This is a great way of optimizing Angular applications which use NgRx.
+
+3. Make repeated calls inexpensive:
+
+   For demonstration purposes we want to follow a different path. We want to memoize the result ourselves and by that assure that the expensive calculation is lazily evaluated and cached for further references.
```

Comment:
"for demo purposes" sounds a bit like "we just do it now because it's easy, but don't do this in production"